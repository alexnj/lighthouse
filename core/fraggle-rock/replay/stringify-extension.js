/**
 * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

import * as PuppeteerReplay from '@puppeteer/replay';

/**
 * @param {PuppeteerReplay.Schema.Step} step
 * @return {boolean}
 */
function isNavigationStep(step) {
  return Boolean(
    step.type === 'navigate' ||
    step.assertedEvents?.some(event => event.type === 'navigation')
  );
}

class LighthouseStringifyExtension extends PuppeteerReplay.PuppeteerStringifyExtension {
  #isProcessingTimespan = false;

  /**
   * @override
   * @param {PuppeteerReplay.LineWriter} out
   * @param {PuppeteerReplay.Schema.UserFlow} flow
   */
  async beforeAllSteps(out, flow) {
    out.appendLine(`const fs = require('fs');`);

    let isMobile = true;
    for (const step of flow.steps) {
      if (step.type !== 'setViewport') continue;
      isMobile = step.isMobile;
    }

    await super.beforeAllSteps(out, flow);

    const flags = {
      screenEmulation: {
        disabled: true,
      },
    };
    out.appendLine(`const flags = ${JSON.stringify(flags)}`);
    if (isMobile) {
      out.appendLine(`const config = undefined;`);
    } else {
      // eslint-disable-next-line max-len
      out.appendLine(`const config = (await import('lighthouse/core/config/desktop-config.js')).default;`);
    }

    out.appendLine(`const lhApi = await import('lighthouse/core/fraggle-rock/api.js');`);
    // eslint-disable-next-line max-len
    out.appendLine(`const lhFlow = await lhApi.startFlow(page, {name: ${JSON.stringify(flow.title)}, config, flags});`);
  }

  /**
   * @override
   * @param {PuppeteerReplay.LineWriter} out
   * @param {PuppeteerReplay.Schema.Step} step
   * @param {PuppeteerReplay.Schema.UserFlow} flow
   */
  async stringifyStep(out, step, flow) {
    if (step.type === 'setViewport') {
      await super.stringifyStep(out, step, flow);
      return;
    }

    const isNavigation = isNavigationStep(step);

    if (isNavigation) {
      if (this.#isProcessingTimespan) {
        out.appendLine(`await lhFlow.endTimespan();`);
        this.#isProcessingTimespan = false;
      }
      out.appendLine(`await lhFlow.startNavigation();`);
    } else if (!this.#isProcessingTimespan) {
      out.appendLine(`await lhFlow.startTimespan();`);
      this.#isProcessingTimespan = true;
    }

    await super.stringifyStep(out, step, flow);

    if (isNavigation) {
      out.appendLine(`await lhFlow.endNavigation();`);
    }
  }

  /**
   * @override
   * @param {PuppeteerReplay.LineWriter} out
   * @param {PuppeteerReplay.Schema.UserFlow} flow
   */
  async afterAllSteps(out, flow) {
    if (this.#isProcessingTimespan) {
      out.appendLine(`await lhFlow.endTimespan();`);
    }
    out.appendLine(`const lhFlowReport = await lhFlow.generateReport();`);
    out.appendLine(`fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)`);
    await super.afterAllSteps(out, flow);
  }
}

export default LighthouseStringifyExtension;
