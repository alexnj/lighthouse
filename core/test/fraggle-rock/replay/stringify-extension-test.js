/**
 * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

import {stringify} from '@puppeteer/replay';

import LighthouseStringifyExtension from '../../../fraggle-rock/replay/stringify-extension.js';

describe('LighthouseStringifyExtension', () => {
  it('handles ending timespan', async () => {
    /** @type {import('@puppeteer/replay').Schema.UserFlow} */
    const flowJson = {
      title: 'Test Flow',
      steps: [
        {
          type: 'setViewport',
          width: 757,
          height: 988,
          deviceScaleFactor: 3,
          isMobile: true,
          hasTouch: true,
          isLandscape: false,
        },
        {
          type: 'navigate',
          url: 'https://example.com',
          assertedEvents: [
            {
              type: 'navigation',
              url: 'https://example.com',
              title: '',
            },
          ],
        },
        {
          type: 'click',
          target: 'main',
          selectors: [['#button']],
          offsetY: 13.5625,
          offsetX: 61,
        },
      ],
    };

    const scriptContents = await stringify(flowJson, {
      extension: new LighthouseStringifyExtension(),
    });

    // Trim the output to the relevant stuff
    const endIndex = scriptContents.indexOf('browser.close');
    const relevantOutput = scriptContents.substring(0, endIndex);

    expect(relevantOutput).toMatchSnapshot();
  });

  it('handles ending navigation', async () => {
    /** @type {import('@puppeteer/replay').Schema.UserFlow} */
    const flowJson = {
      title: 'Test Flow',
      steps: [
        {
          type: 'setViewport',
          width: 757,
          height: 988,
          deviceScaleFactor: 3,
          isMobile: true,
          hasTouch: true,
          isLandscape: false,
        },
        {
          type: 'navigate',
          url: 'https://example.com',
          assertedEvents: [
            {
              type: 'navigation',
              url: 'https://example.com',
              title: '',
            },
          ],
        },
        {
          type: 'click',
          target: 'main',
          selectors: [['#button']],
          offsetY: 13.5625,
          offsetX: 61,
        },
        {
          type: 'navigate',
          url: 'https://example.com/page/',
          assertedEvents: [
            {
              type: 'navigation',
              url: 'https://example.com/page/',
              title: '',
            },
          ],
        },
      ],
    };

    const scriptContents = await stringify(flowJson, {
      extension: new LighthouseStringifyExtension(),
    });

    // Trim the output to the relevant stuff
    const endIndex = scriptContents.indexOf('browser.close');
    const relevantOutput = scriptContents.substring(0, endIndex);

    expect(relevantOutput).toMatchSnapshot();
  });

  it('handles multiple sequential navigations', async () => {
    /** @type {import('@puppeteer/replay').Schema.UserFlow} */
    const flowJson = {
      title: 'Test Flow',
      steps: [
        {
          type: 'setViewport',
          width: 757,
          height: 988,
          deviceScaleFactor: 3,
          isMobile: true,
          hasTouch: true,
          isLandscape: false,
        },
        {
          type: 'navigate',
          url: 'https://example.com',
          assertedEvents: [
            {
              type: 'navigation',
              url: 'https://example.com',
              title: '',
            },
          ],
        },
        {
          type: 'click',
          target: 'main',
          selectors: [['#link']],
          offsetY: 13.5625,
          offsetX: 61,
          assertedEvents: [
            {
              type: 'navigation',
              url: 'https://example.com/page',
              title: '',
            },
          ],
        },
      ],
    };

    const scriptContents = await stringify(flowJson, {
      extension: new LighthouseStringifyExtension(),
    });

    // Trim the output to the relevant stuff
    const endIndex = scriptContents.indexOf('browser.close');
    const relevantOutput = scriptContents.substring(0, endIndex);

    expect(relevantOutput).toMatchSnapshot();
  });
});
