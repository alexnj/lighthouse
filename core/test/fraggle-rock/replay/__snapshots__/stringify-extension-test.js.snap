// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LighthouseStringifyExtension handles ending navigation 1`] = `
"const fs = require('fs');
const puppeteer = require('puppeteer'); // v13.0.0 or later

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  const timeout = 5000;
  page.setDefaultTimeout(timeout);

  const flags = {\\"screenEmulation\\":{\\"disabled\\":true}}
  const config = undefined;
  const lhApi = await import('lighthouse/core/fraggle-rock/api.js');
  const lhFlow = await lhApi.startFlow(page, {name: \\"Test Flow\\", config, flags});
  {
    const targetPage = page;
    await targetPage.setViewport({\\"width\\":757,\\"height\\":988})
  }
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"https://example.com\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"#button\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 61,
        y: 13.5625,
      },
    });
  }
  await lhFlow.endTimespan();
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"https://example.com/page/\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  const lhFlowReport = await lhFlow.generateReport();
  fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)

  await "
`;

exports[`LighthouseStringifyExtension handles ending timespan 1`] = `
"const fs = require('fs');
const puppeteer = require('puppeteer'); // v13.0.0 or later

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  const timeout = 5000;
  page.setDefaultTimeout(timeout);

  const flags = {\\"screenEmulation\\":{\\"disabled\\":true}}
  const config = undefined;
  const lhApi = await import('lighthouse/core/fraggle-rock/api.js');
  const lhFlow = await lhApi.startFlow(page, {name: \\"Test Flow\\", config, flags});
  {
    const targetPage = page;
    await targetPage.setViewport({\\"width\\":757,\\"height\\":988})
  }
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"https://example.com\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"#button\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 61,
        y: 13.5625,
      },
    });
  }
  await lhFlow.endTimespan();
  const lhFlowReport = await lhFlow.generateReport();
  fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)

  await "
`;

exports[`LighthouseStringifyExtension handles multiple sequential navigations 1`] = `
"const fs = require('fs');
const puppeteer = require('puppeteer'); // v13.0.0 or later

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  const timeout = 5000;
  page.setDefaultTimeout(timeout);

  const flags = {\\"screenEmulation\\":{\\"disabled\\":true}}
  const config = undefined;
  const lhApi = await import('lighthouse/core/fraggle-rock/api.js');
  const lhFlow = await lhApi.startFlow(page, {name: \\"Test Flow\\", config, flags});
  {
    const targetPage = page;
    await targetPage.setViewport({\\"width\\":757,\\"height\\":988})
  }
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"https://example.com\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    const element = await waitForSelectors([[\\"#link\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 61,
        y: 13.5625,
      },
    });
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  const lhFlowReport = await lhFlow.generateReport();
  fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)

  await "
`;
